services/datamanager/builtin/sync/file_deletion.go

```diff
@@ -0,0 +1,260 @@
+			maybeDeleteExcessFiles(ctx, fileTracker, captureDir, deleteEveryNth, clock, logger)
+		}
+	}
+}
+
+func maybeDeleteExcessFiles(
+	ctx context.Context,
+	fileTracker *fileTracker,
+	captureDir string,
+	deleteEveryNth int,
+	clock clock.Clock,
+	logger logging.Logger,
+) {
+	start := clock.Now()
+	logger.Debug("checking disk usage")
+	usage, err := diskusage.Statfs(captureDir)
+	logger.Debugf("disk usage: %s", usage)
+	if err != nil {
+		logger.Error(errors.Wrap(err, "error checking file system stats"))
+		return
+	}
+
+	if usage.SizeBytes == 0 {
+		logger.Error("captureDir partition has size zero")
+		return
+	}
+	deletedFileCount, err := deleteExcessFiles(
+		ctx,
+		fileTracker,
+		usage,
+		captureDir,
+		deleteEveryNth,
+		logger)
+
+	duration := clock.Since(start)
+
+	switch {
+	case err != nil:
+		logger.Errorw("error deleting cached datacapture files", "error", err, "execution time", duration.Seconds())
+	case deletedFileCount > 0:
+		logger.Infof("%d files have been deleted to avoid the disk filling up, execution time: %f", deletedFileCount, duration.Seconds())
+	default:
+		logger.Infof("no files delete, execution time: %s", duration)
+	}
+}
+
+func deleteExcessFiles(
+	ctx context.Context,
+	fileTracker *fileTracker,
+	usage diskusage.DiskUsage,
+	captureDir string,
+	deleteEveryNth int,
+	logger logging.Logger,
+) (int, error) {
+	shouldDelete, err := shouldDeleteBasedOnDiskUsage(
+		ctx,
+		usage,
+		captureDir,
+		logger)
+	if err != nil {
+		return 0, errors.Wrap(err, "error checking file system stats")
+	}
+
+	if !shouldDelete {
+		return 0, nil
+	}
+
+	logger.Warnf("current disk usage of the data capture directory exceeds threshold (%f)", CaptureDirToFSUsageRatio)
+	return deleteFiles(ctx, fileTracker, deleteEveryNth, captureDir, logger)
+}
+
+func shouldDeleteBasedOnDiskUsage(
+	ctx context.Context,
+	usage diskusage.DiskUsage,
+	captureDirPath string,
+	logger logging.Logger,
+) (bool, error) {
+	usedSpace := 1.0 - usage.AvailablePercent()
+	if usedSpace < FSThresholdToTriggerDeletion {
+		logger.Debugf("disk not full enough. Threshold: %s, Used space: %s, %s",
+			fmt.Sprintf("%.2f", FSThresholdToTriggerDeletion*100)+"%",
+			fmt.Sprintf("%.2f", usedSpace*100)+"%",
+			usage)
+		return false, nil
+	}
+	// Walk the dir to get capture stats
+	shouldDelete, err := exceedsDeletionThreshold(
+		ctx,
+		captureDirPath,
+		float64(usage.SizeBytes),
+		CaptureDirToFSUsageRatio,
+	)
+	if err != nil && !shouldDelete {
+		logger.Warnf("Disk nearing capacity but data capture directory is below %f of that size, file deletion will not run",
+			CaptureDirToFSUsageRatio)
+	}
+	return shouldDelete, err
+}
+
+// returns false, nil if the threshold is not exceeded
+// returns false, error if an IO error was encountered
+// returns true, nil if the threshold is exceeded.
+func exceedsDeletionThreshold(
+	ctx context.Context,
+	captureDirPath string,
+	fileSystemSizeBytes float64,
+	captureDirToFSUsageRatio float64,
+) (bool, error) {
```

Comment:
suggest using named return values for clarity
```suggestion
) (exceeds bool, err error) {
```

Reply:
I try to avoid named returns especially for functions this long as I think they make it easy to return values you didn't intend to mistakenly.



from: https://go.dev/tour/basics/7 
```
Named return values
Go's return values may be named. If so, they are treated as variables defined at the top of the function.

These names should be used to document the meaning of the return values.

A return statement without arguments returns the named return values. This is known as a "naked" return.

Naked return statements should be used only in short functions, as with the example shown here. They can harm readability in longer functions.
```